axelrod/strategies/dbs.py

```diff
@@ -0,0 +1,354 @@
+    tection for detecting whether anomalies in player’s behavior are deliberate or accidental, hence performing quite well in noisy tournaments.  
+
+    From the learned opponent's strategy, a tree search is used to choose the best move
+
+
+
+
+    Default values for the parameters are the suggested values in the article 
+    When more noise you can try to diminish violation_threshold and rejection_threshold
+
+    Parameters:
+
+    discount_factor: (float, between 0 and 1) discount factor used when computing discounted frequencies to learn opponent's strategy
+            default: .75
+
+    promotion_threshold: (int) number of observations needed to promote a change in opponent's strategy
+            default: 3
+
+    violation_threshold: (int) number of observation needed to considerate opponent's strategy has changed. Seems good to lower it when noise increase
+            default: 4
+
+    reject_threshold: (int) number of observations before forgetting opponents old strategy. Seems good to lower it when noise increase
+            default: 3
+
+    tree_depth: (int) depth of the tree for the tree-search algorithm. 
+            default is 5.
+            Higher depth means more time to coompute the move.
+
+    """
+
+    # These are various properties for the strategy
+    # Not accurately precised yet
+    name = 'DBS'
+    classifier = {
+        'memory_depth': float('inf'),
+        'stochastic': False,
+        'makes_use_of': set(),
+        'long_run_time': False,
+        'inspects_source': False,
+        'manipulates_source': False,
+        'manipulates_state': False
+    }
+
+    
+    # best value for reject_threshold ? not mentioned in the article
+    def __init__(self, discount_factor=.75, promotion_threshold=3, violation_threshold=4, reject_threshold=3,tree_depth=5): 
+        super().__init__()
+
+        # default opponent's policy is TitForTat
+        self.Rd = Policy.prob_policy(1,1,0,0)
+        self.Rc = Policy()
+        self.Pi = self.Rd   # policy used by MoveGen
+        self.violation_counts = Policy()
+        self.reject_threshold = reject_threshold
+        self.violation_threshold = violation_threshold
+        self.promotion_threshold = promotion_threshold
+        self.tree_depth = tree_depth
+        self.v = 0
+        self.alpha = discount_factor
+        self.history_by_cond = {}
+        # to compute the discount frequencies, we need to keep
+        # up to date an history of what has been played for each
+        # condition:
+        # We save it as a dict history_by_cond; keys are conditions (ex (C,C)) 
+        # and values are a tuple of 2 lists (G,F)
+        # for a condition j: 
+        # G[i] = 1 if cond j was True at turn i-1 and C has been played
+        # by the opponent; else G[i]=0
+        # F[i] = 1 if cond j was True at turn i-1; else G[i]=0
+        # initial hypothesized policy is TitForTat
+        self.history_by_cond[(C,C)]=([1],[1])
```

Comment:
We use PEP8's recommended formatting. For this line it would be:
`self.history_by_cond[(C, C)] = ([1], [1])`
(and similarly below)